swift/ql/src/queries/Security/CWE-321/HardcodedEncryptionKey.ql

```diff
@@ -0,0 +1,70 @@
+/**
+ * @name Hard-coded encryption key
+ * @description Using hardcoded keys for encryption is not secure, because potential attacker can easiy guess them.
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 8.1
+ * @precision high
+ * @id swift/hardcoded-key
+ * @tags security
+ *       external/cwe/cwe-321
+ */
+
+import swift
+import codeql.swift.dataflow.DataFlow
+import DataFlow::PathGraph
+
+/**
+ * An `Expr` that is used to initialize a key.
+ */
+abstract class KeySource extends Expr { }
+
+/**
+ * The creation of a literal byte array.
+ */
+class ByteArrayLiteralSource extends KeySource {
+  ByteArrayLiteralSource() { this = any(ArrayExpr arr | arr.getType().toString() = "Array<UInt8>") }
+}
+
+/**
+ * Any string literal as a key source.
+ */
+class StringLiteralSource extends KeySource {
+  StringLiteralSource() { this instanceof StringLiteralExpr }
+}
+
+/**
+ * A class for all ways to set a key.
+ */
+class EncryptionKeySink extends Expr {
+  EncryptionKeySink() {
+    // `key` arg in `init` is a sink
+    exists(ClassDecl c, AbstractFunctionDecl f, CallExpr call |
```

Comment:
Should this `ClassDecl` be a `ClassOrStructDecl`? Or do we know that they're all actually `class`es?

Reply:
They are indeed all classes.